### ISSUE TITLE
Distributor failing with 500s for no clear reason

### ISSUE THREAD

**Issue by jakubgs** [cortexproject/cortex#4158]
**Describe the bug**
I'm seeing random 500s when Prometheus is pushing metrics to `/api/v1/push`:
```
msg="Failed to send batch, retrying" err="server returned HTTP status 500 Internal Server Error: rpc error: code = Unavailable desc = transport is closing"
```
Which looks like this on the Cortex side:
```
msg="POST /api/v1/push (500) 281.388644ms Response: \"rpc error: code = Unavailable desc = transport is closing\\n\" ws: false; Content-Encoding: snappy; Content-Length: 32409; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
```
But it's just a `warn` level message, and even with `debug` logs I see no reason for this error.

The number of samples being sent is tiny:
```
 > curl -s localhost:9090/metrics | grep "^prometheus_tsdb_head_series "
prometheus_tsdb_head_series 34294
```
And the hosts are VERY beefy and underutilized, so I'm really confused why this is happening

**To Reproduce**
Not really sure. I'm happy to help debug this, but I'm not sure where to start.

**Expected behavior**
Error should include reason for 500 error, but all it contains is `rpc error: code = Unavailable desc = transport is closing`.

**Environment:**
 - Infrastructure: Systemd service on Ubuntu
 - Version: `1.8.0`

**Storage Engine**
Chunks storage using Cassandra `3.11.9`.

**Additional Context**
I started getting a LOT of 500s suddenly, so I disabled all Prometheus instances except one to debug this, but the logs give me no indication as to why it's actually happening. When I re-enable all other Prometheus instances the 500s keep raising until they overwhelm Cortex.
![image](https://user-images.githubusercontent.com/2212681/117213869-761e6780-adfc-11eb-919e-4ce382c41cd4.png)

**Comment by jakubgs** [comment]
One thing that confuses me is why the error mentions RPC, when the request is a normal HTTP post.
This suggests to me that the issue is in grpc connection between Distributor and Ingester, possibly.

I found these two issues:
- https://github.com/grpc/grpc-go/issues/4040
- https://github.com/grpc/grpc-go/issues/2160

Where the second one leads to this StackOverflow suggestion:

>My solution turned out to be fairly simple:
>```go
>server = grpc.NewServer(
>    grpc.KeepaliveParams(keepalive.ServerParameters{
>        MaxConnectionIdle: 5 * time.Minute,           // <--- This fixes it!
>    }),
>)
>```
>This ensures that the gRPC server closes the underlying TCP socket gracefully itself before any abrupt kills from the kernel or intermediary servers (AWS and Google Cloud Load Balancers both have larger timeouts than 5 minutes).

https://stackoverflow.com/a/54703234

But not sure if that's related.

**Comment by jakubgs** [comment]
These appear to be the GRCP defaults for related settings:
```yaml
# The duration after which an idle connection should be closed. Default:
# infinity
# CLI flag: -server.grpc.keepalive.max-connection-idle
[grpc_server_max_connection_idle: <duration> | default = 2562047h47m16.854775807s]

# The duration for the maximum amount of time a connection may exist before it
# will be closed. Default: infinity
# CLI flag: -server.grpc.keepalive.max-connection-age
[grpc_server_max_connection_age: <duration> | default = 2562047h47m16.854775807s]

# An additive period after max-connection-age after which the connection will be
# forcibly closed. Default: infinity
# CLI flag: -server.grpc.keepalive.max-connection-age-grace
[grpc_server_max_connection_age_grace: <duration> | default = 2562047h47m16.854775807s]
```
Which essentially amounts to "forever". Not sure if that's some kind of optimization.

**Comment by jakubgs** [comment]
Oh, I see something more:
```
caller=frontend_processor.go:59 msg="error processing requests" address=10.1.31.152:9095 err="rpc error: code = Unavailable desc = transport is closing"
```
Which appears to be coming from here:
https://github.com/cortexproject/cortex/blob/51662ea3e57cb2639cc057f5b647e8bc671ff282/pkg/querier/worker/frontend_processor.go#L58-L62

**Comment by jakubgs** [comment]
No, I think that's a red-herring, or a related but separate issue, and is caused by my query frontend which shows:
```
msg="error processing request" try=0 err="rpc error: code = Code(500) desc = {\"status\":\"error\",\"errorType\":\"internal\",\"error\":\"expanding series: rpc error: code = Canceled desc = grpc: the client connection is closing\"}"
```
Not sure what that's about either tho...

**Comment by pstibrany** [comment]
Internally Cortex uses gRPC to communicate between various components, even if those components run inside single binary.

If your Cortex cluster has very low usage, you may be running into problem described here: https://github.com/cortexproject/cortex/issues/3606#issuecomment-745179803

**Comment by jakubgs** [comment]
Well, I tried to do low number of samples as a way to debug this. Normally we have a bit under 1 million samples.

But when i leave it be the number of 500s just keeps raising. No idea why:
![](https://user-images.githubusercontent.com/2212681/117262337-6b41f200-ae51-11eb-9091-a01112e0c1d0.png)

I'm really tearing my hair out here. It's kinda absurd when I think about it. A single legacy Prometheus federating host handles all of those metrics fine with 8 GB of ram and 4 cores, while a 3x8core and 16 GB ram Cortex cluster struggles to handle the same amount. I'm __assuming__ it's a misconfiguration issue, but there's so many options one could spend a lifetime randomly tweaking them one at a time and not find the right setup.

**Comment by jakubgs** [comment]
Although the 500s I'm seeing now are caused by:
```
rpc error: code = DeadlineExceeded desc = context deadline exceeded
```
So I guess that's a separate issue form the one I saw when I tried running just one Prometheus instance feeding Cortex.

Not a very helpful error tho. Not sure where it's coming from, what context deadline is exceeded, since `caller` is `logging.go:71`.

**Comment by jakubgs** [comment]
Now I'm seeing a lot of:
```
msg="removing ingester failing healthcheck" addr=10.1.1.X:9095 reason="failing healthcheck status: NOT_SERVING"
```
And based on this function:
https://github.com/cortexproject/cortex/blob/51662ea3e57cb2639cc057f5b647e8bc671ff282/pkg/util/grpc/healthcheck/health_check.go#L40-L53

I deduce that the issue is that not all services are started. And indeed:
```
 > curl -s 'http://localhost:9092/services' | awk -F'</?td>' '/td/{name=$2; getline; printf "%30s - %s\n", name, $2}'
                 chunks-purger - Starting
           distributor-service - Running
              ingester-service - Running
                 memberlist-kv - Running
                       querier - Running
                query-frontend - Running
    query-frontend-tripperware - Running
                          ring - Running
                        server - Running
                         store - Running
                 table-manager - Runnin
```
It appears that `chunks-purger` is endlessly stuck in `Starting` state...

**Comment by jakubgs** [comment]
Truncating the `delete_requests` table fixed it:
```
[cqlsh 5.0.1 | Cassandra 3.11.9 | CQL spec 3.4.4 | Native protocol v4]
Use HELP for help.
admin@cqlsh> USE cortex_metrics;
admin@cqlsh:cortex_metrics> TRUNCATE delete_requests;
```
And now all services appear to be `Running:
```
 > curl -s 'http://localhost:9092/services' | awk -F'</?td>' '/td/{name=$2; getline; printf "%30s - %s\n", name, $2}'
                 chunks-purger - Running
           distributor-service - Running
              ingester-service - Running
                 memberlist-kv - Running
                       querier - Running
                query-frontend - Running
    query-frontend-tripperware - Running
                          ring - Running
                        server - Running
                         store - Running
                 table-manager - Running
```
It appears that the `chunks-purger` was failing because of errors like:
```
caller=purger.go:415 user_id=fake request_id=c637975f msg="failed to resume stalled request" err="failed to build delete plan: session has been closed"
caller=cortex.go:424 msg="module failed" module=chunks-purger err="context canceled
caller=purger.go:415 user_id=fake request_id=c637975f msg="failed to resume stalled request" err="failed to build delete plan: gocql: connection closed waiting for response
caller=cortex.go:424 msg="module failed" module=chunks-purger err="context canceled"
```
It's always something.

**Comment by jakubgs** [comment]
It makes no sense. It works perfectly fine for hours and then suddenly a huge spike in 500s:

![image](https://user-images.githubusercontent.com/2212681/117432774-00fd8000-af2b-11eb-9210-74690ca9ac42.png)

**Comment by jakubgs** [comment]
One of the hosts in the cluster, I assume the ingester ring master, suddenly starts reporting a huge surge in samples:

![image](https://user-images.githubusercontent.com/2212681/117432848-1a063100-af2b-11eb-8e1c-facb6038db21.png)

**Comment by jakubgs** [comment]
And Cassandra write latency is totally fine the whole time:

![image](https://user-images.githubusercontent.com/2212681/117432976-3e620d80-af2b-11eb-8017-c8992776bc95.png)

**Comment by jakubgs** [comment]
It's just an endless stream of 500s clearly not caused by Cassandra storage backend:
```
msg="POST /api/v1/push (500) 2.330206842s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 43105; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
msg="POST /api/v1/push (500) 2.069391565s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 43000; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
msg="POST /api/v1/push (500) 2.104685298s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 42431; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
msg="POST /api/v1/push (500) 2.097729202s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 41756; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
msg="POST /api/v1/push (500) 2.074144232s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 40431; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
msg="POST /api/v1/push (500) 2.096689061s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 41995; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
msg="POST /api/v1/push (500) 2.109004699s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 39052; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
msg="POST /api/v1/push (500) 2.382432777s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 43313; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
msg="POST /api/v1/push (500) 2.954348261s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 42914; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
msg="POST /api/v1/push (500) 2.70520163s Response: \"rpc error: code = DeadlineExceeded desc = context deadline exceeded\\n\" ws: false; Content-Encoding: snappy; Content-Length: 44273; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.26.0; X-Prometheus-Remote-Write-Version: 0.1.0; 
```

**Comment by jakubgs** [comment]
Interesting, I'm seeing a big TCP sockets counts raise when the 500 start happening:
![image](https://user-images.githubusercontent.com/2212681/117453992-e84e9380-af45-11eb-844e-a6e427cd6744.png)

**Comment by jakubgs** [comment]
And they are clearly from Cortex:
```
 > sudo netstat -tupen | grep -e ESTABLISHED -e cortex | wc -l   
1097
```

**Comment by jakubgs** [comment]
And those connections are on the HTTP port, and not the GRPC port:
```
admin@master-01.do-ams3.metrics.hq:~ % sudo netstat -tupen | grep -e ESTABLISHED -e cortex -e 9092 | wc -l
1025
```
So they are from Prometheus.

**Comment by jakubgs** [comment]
I tried lowering `max_shards` on Prometheus, since I thought that might be the reason for the 500s, but it's definitely not:

![image](https://user-images.githubusercontent.com/2212681/117480817-5e60f380-af62-11eb-854b-da955b54be35.png)

I really don't get how I'm supposed to debug this. All I'm getting is `context deadline exceeded`, and cortex shows no errors, aside from the usual `sample timestamp out of order` ones.

For example this `cortex_distributor_ingester_append_failures_total` metric:
![image](https://user-images.githubusercontent.com/2212681/117480995-a41dbc00-af62-11eb-892f-20808bee17d5.png)

Why is it going up? How can I find out the reason?

**Comment by jakubgs** [comment]
According to Prometheus metrics the affected Cortex host doesn't receive any samples:

![image](https://user-images.githubusercontent.com/2212681/117483358-ea284f00-af65-11eb-8169-8440ad9a4858.png)

**Comment by jakubgs** [comment]
As far as I can tell the `context deadline exceeded` error comes from this line:
https://github.com/cortexproject/cortex/blob/51662ea3e57cb2639cc057f5b647e8bc671ff282/pkg/distributor/distributor.go#L681

**Comment by jakubgs** [comment]
Based on increase in `cortex_ingester_ingested_samples_failures_total` I deduce that the error comes from:
https://github.com/cortexproject/cortex/blob/51662ea3e57cb2639cc057f5b647e8bc671ff282/pkg/ingester/ingester.go#L479-L480

**Comment by jakubgs** [comment]
Just to clarify, I have tried adjusting GRPC ping settings and they look like this:
```yaml
server:
  grpc_server_keepalive_timeout: 120s
  grpc_server_min_time_between_pings: 5s
  grpc_server_ping_without_stream_allowed: true
```
But I'm still seeing a constant stream of `context deadline exceeded` errors. There's usually a burst of 5-10 every few seconds.

**Comment by jakubgs** [comment]
I printed the context for the `Push` calls and it looks like this:
```
context.Background.WithDeadline(2021-05-10 08:21:08.364624619 +0000 UTC m=+41.680762449 [1.99987099s]).WithValue(type user.contextKey, val fake).WithValue(type opentracing.contextKey, val 4a53e95c88c92768:4a53e95c88c92768:0000000000000000:0)
```
The time limit is `1.99987099s`, essentiall 2 seconds, which matches the `distributor.remote_timeout` default:
```yaml
# Timeout for downstream ingesters.
# CLI flag: -distributor.remote-timeout
[remote_timeout: <duration> | default = 2s]
```
Which I assume is the source of this timeout value. But what is actually causing the timeout?

**Comment by jakubgs** [comment]
I've increased `distributor.remote_timeout` to `4s` and it indeed has changed this context timeout:
```
context.Background.WithDeadline(2021-05-10 09:33:05.779113643 +0000 UTC m=+542.676683704 [3.999978785s]).WithValue(type user.contextKey, val fake).WithValue(type opentracing.contextKey, val 1f64a945e4fb6cb6:1f64a945e4fb6cb6:0000000000000000:0)
```

**Comment by jakubgs** [comment]
I'm not 100% if this exactly what helped, but I disabled `ingester.walconfig.wal_enabled`, and currently it's looking healthy.

I also lowering `ingester.max_chunk_age` to 2 hours and adjusted Memcached config to use `batch_size: 16384` and `parallelism: 10` for `chunk_store` and `storage`.

**Comment by jakubgs** [comment]
I've created the certificates using `certbot` in a Docker container and yet I'm still getting this error:
```
 EOF occurred in violation of protocol (_ssl.c:1131)
```
It's always something...

**Comment by jakubgs** [comment]
Somethings I've tried and might have helped, separately or in combination.

* I've replaced our userland [Tinc VPN](https://www.tinc-vpn.org/) with built into kernel [WireGuard VPN](https://www.wireguard.com/)
  - This definitely reduced latency and increased bandwidth between services, and reduced CPU usage.
* I moved the Memcached service to a dedicated host with more memory
  - I'm not running a cluster, but just one node for now, but it seems to be enough for our use.
* Separated distributors from ingesters, which before were running as part of `all` target
  - This seems to have helped with handling of cortex restarts, since distributors can handle selecting different ingester better.

After making all these changes I no longer see random 500s without any clear reason. I'm not sure what contributed the most.

My best guess would be that the presence of distributors and ingesters on the same host somehow causes GRPC timeouts.
